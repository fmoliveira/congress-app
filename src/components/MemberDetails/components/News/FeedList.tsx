import React from "react"

import { FeedItem } from "./FeedItem"

interface IProps {
  feed: any[]
}

export const FeedList = ({ feed = [] }: IProps) => (
  <div>
    {feed.map(i => (
      <FeedItem key={i.key} feed={i} />
    ))}
  </div>
)
