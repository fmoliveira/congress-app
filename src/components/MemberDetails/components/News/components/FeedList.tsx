import React from "react"
import shortId from "shortid"

import { IFeedItem } from "../types"
import { FeedItem } from "./index"

interface IProps {
  feed: IFeedItem[]
}

const FeedList = ({ feed = [] }: IProps) => (
  <div>
    {feed.map(i => (
      <FeedItem key={shortId.generate()} feed={i} />
    ))}
  </div>
)

export default FeedList
