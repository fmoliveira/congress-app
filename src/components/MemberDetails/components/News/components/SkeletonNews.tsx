import React from "react"

import { TextSkeleton } from "../../../../_common"

const SKELETON_ITEMS_AMOUNT = 5

const items = new Array(SKELETON_ITEMS_AMOUNT).fill(0).map((_, idx) => idx)

export const SkeletonNews = () => (
  <div>
    {items.map(i => (
      <TextSkeleton key={i} fullWidth={true}>
        Loading
      </TextSkeleton>
    ))}
  </div>
)
