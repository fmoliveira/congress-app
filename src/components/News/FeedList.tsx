import React from "react"

interface IProps {
  feed: any[]
}

export const FeedList = ({ feed = [] }: IProps) => (
  <div>{feed.length} items</div>
)
