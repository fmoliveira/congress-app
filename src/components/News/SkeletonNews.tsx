import React from "react"

import { TextSkeleton } from "../LoadingSkeleton"

const SKELETON_ITEMS_AMOUNT = 5

const items = new Array(SKELETON_ITEMS_AMOUNT).fill(0).map((_, idx) => idx)

export const SkeletonNews = () => (
  <div>
    {items.map(i => (
      <TextSkeleton key={i} fullWidth={true}>
        Loading
      </TextSkeleton>
    ))}
  </div>
)
