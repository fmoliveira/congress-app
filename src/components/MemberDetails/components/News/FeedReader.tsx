import React, { PureComponent } from "react"
import { connect } from "react-redux"

import { noop } from "lodash"

import { RequestStatusType } from "../../../../reducers/requestStatus"
import { loadNewsFeed } from "./actions"
import {
  loadNewsFeedStatusSelector,
  newsFeedByMemberIdSelector
} from "./reducers"
import { IFeedItem } from "./types"

import { FeedList, SkeletonNews } from "./components"

interface IOwnProps {
  memberId: string
  feedUrl: string
}

interface IStateProps {
  feed: IFeedItem[]
  status: RequestStatusType
}

interface IDispatchProps {
  loadNewsFeed: (memberId: string, feedUrl: string) => void
}

type Props = IOwnProps & IStateProps & IDispatchProps

class FeedReader extends PureComponent<Props> {
  public static defaultProps = {
    loadNewsFeed: noop
  }

  public componentDidMount() {
    this.loadNewsFeed()
  }

  public render() {
    const { feed, status } = this.props

    return (
      <div>
        {status === RequestStatusType.Loading && <SkeletonNews />}
        {status === RequestStatusType.Error && (
          <div>Failed to load news feed!</div>
        )}
        {status === RequestStatusType.Success && <FeedList feed={feed} />}
      </div>
    )
  }

  private loadNewsFeed() {
    const { memberId, feedUrl } = this.props

    if (feedUrl) {
      this.props.loadNewsFeed(memberId, feedUrl)
    }
  }
}

const mapStateToProps = (state: any, props: any) => ({
  feed: newsFeedByMemberIdSelector(state, props.memberId),
  status: loadNewsFeedStatusSelector(state)
})

const mapDispatchToProps = {
  loadNewsFeed
}

export { FeedReader as UnconnectedFeedReader }

export default connect<IStateProps, IDispatchProps>(
  mapStateToProps,
  mapDispatchToProps
)(FeedReader)
