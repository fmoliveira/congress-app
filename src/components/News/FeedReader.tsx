import React, { PureComponent } from "react"
import { connect } from "react-redux"

import { RequestStatusType } from "../../reducers/requestStatus"
import { loadNewsFeed } from "./actions"
import {
  loadNewsFeedStatusSelector,
  newsFeedByMemberIdSelector
} from "./reducers"

import { FeedList } from "./FeedList"
import { SkeletonNews } from "./SkeletonNews"

interface IOwnProps {
  memberId: string
  feedUrl: string
}

interface IStateProps {
  feed: any[]
  status: RequestStatusType
}

interface IDispatchProps {
  loadNewsFeed: (memberId: string, feedUrl: string) => void
}

type Props = IOwnProps & IStateProps & IDispatchProps

class FeedReader extends PureComponent<Props> {
  public componentDidMount() {
    this.loadNewsFeed()
  }

  public render() {
    const { feed, status } = this.props

    return (
      <div>
        {status === RequestStatusType.Loading && <SkeletonNews />}
        {status === RequestStatusType.Success && <FeedList feed={feed} />}
      </div>
    )
  }

  private loadNewsFeed() {
    const { memberId, feedUrl } = this.props

    if (feedUrl) {
      this.props.loadNewsFeed(memberId, feedUrl)
    }
  }
}

const mapStateToProps = (state: any, props: any) => ({
  feed: newsFeedByMemberIdSelector(state, props.memberId),
  status: loadNewsFeedStatusSelector(state)
})

const mapDispatchToProps = {
  loadNewsFeed
}

export default connect<IStateProps, IDispatchProps>(
  mapStateToProps,
  mapDispatchToProps
)(FeedReader)
