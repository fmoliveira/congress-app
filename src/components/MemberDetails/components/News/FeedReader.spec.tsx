import React from "react"
import { cleanup, render } from "react-testing-library"

import FeedList from "./components/FeedList"
import SkeletonNews from "./components/SkeletonNews"
import { UnconnectedFeedReader as FeedReader } from "./FeedReader"

jest.mock("./components/FeedList", () => jest.fn(() => <div>FeedList</div>))
jest.mock("./components/SkeletonNews", () =>
  jest.fn(() => <div>SkeletonNews</div>)
)

const setup = (props?: any) => render(<FeedReader {...props} />)

describe("FeedReader", () => {
  afterEach(() => {
    cleanup()
    jest.clearAllMocks()
  })

  it("Should render without crashing", () => {
    const { container } = setup()
    expect(container.firstChild).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  it("Should call action to load news feed", () => {
    const loadNewsFeed = jest.fn()
    const props = {
      feedUrl: "https://read.me/@a250/feed.rss",
      loadNewsFeed,
      memberId: "A250"
    }
    setup(props)

    expect(loadNewsFeed).toHaveBeenCalledTimes(1)
    expect(loadNewsFeed).toHaveBeenCalledWith(
      "A250",
      "https://read.me/@a250/feed.rss"
    )
  })

  it("Should render feed list", () => {
    const props = {
      feed: [
        { link: "https://read.me/@a250/first-story", title: "First Story" }
      ],
      status: "SUCCESS"
    }
    setup(props)

    expect(SkeletonNews).not.toHaveBeenCalled()
    expect(FeedList).toHaveBeenCalledTimes(1)
    expect(FeedList).toHaveBeenCalledWith(
      {
        feed: [
          { link: "https://read.me/@a250/first-story", title: "First Story" }
        ]
      },
      {}
    )
  })

  it("Should render skeleton while loading", () => {
    const props = {
      status: "REQUEST"
    }
    setup(props)

    expect(FeedList).not.toHaveBeenCalled()
    expect(SkeletonNews).toHaveBeenCalledTimes(1)
  })
})
