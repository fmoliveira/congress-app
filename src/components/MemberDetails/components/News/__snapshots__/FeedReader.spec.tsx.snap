// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedReader Should render without crashing 1`] = `
<div>
  <div />
</div>
`;
